### ISSUE TITLE
Ad-rendering with client-side cached advertisements is broken

### ISSUE THREAD

**Issue by Caspervw** [prebid/Prebid.js#2225]
## Type of issue
Bug

## Description
The renderAd function in combination with DFP and client-side cached ads is broken. If one ad is cached and another is requested, we can run into the following state:

```
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_format_appnexus value: banner
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_source_appnexus value: client
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_size_appnexus value: 300x250
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_adid_appnexus value: 4fd87aed1b83d,10c8b8b14b81e02
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_bidder_appnexus value: appnexus
MESSAGE: Attempting to set key value for slot: CP-300x250-ATF key: hb_pb_appnexus value: 0.12,0.15
MESSAGE: Emitting event for: setTargeting
INFO: Invoking pbjs.renderAd : params :  Arguments(2) [document, "4fd87aed1b83d,10c8b8b14b81e02", callee: (...), Symbol(Symbol.iterator): ƒ]
MESSAGE: Calling renderAd with adId :4fd87aed1b83d,10c8b8b14b81e02
Error trying to write ad. **Cannot find ad by given id : 4fd87aed1b83d,10c8b8b14b81e02**
```

- The ad with id 4fd87aed1b83d was cached, not rendered and not expired, so was send to DFP.
- The ad with id 10c8b8b14b81e02 was fresh and therefore send to DPF.

DFP can split the hb_pb value into a bid of 0.12 and 0.15 and does do the price-priority competition correct, however the creative can't distinguish values and will return the full hb_adid value.

```javascript
<script>
var w = window;
for (i = 0; i < 10; i++) {
  w = w.parent;
  if (w.pbjs) {
    try {
      w.pbjs.renderAd(document, '%%PATTERN:hb_adid_appnexus%%');
      break;
    } catch (e) {
      continue;
    }
  }
}
</script>
```

This will result in a call to renderAd with adId '4fd87aed1b83d,10c8b8b14b81e02', which does not exist.

## Steps to reproduce
Refresh to get multiple bids. Make sure one's cached and checkout the targeting options for the second DFP-call.

### Expected results
The advertisement with the highest CPM should be shown

### Actual results
Prebid throws an error

## Platform details
- Prebid 1.5.0-pre
- Chrome
- SPA-style application

**Comment by Caspervw** [comment]
Created a pull-request for this issue: https://github.com/prebid/Prebid.js/pull/2226

**Comment by mkendall07** [comment]
@Caspervw 
do you happen to have a test page that demonstrates this?

**Comment by Caspervw** [comment]
@mkendall07 Not a permanent one, but for the next hour I can leave it on production. Keep picking "categories" from "Související kategorie" on http://poki.cz/nákladní-auto

I've loaded up a full development build.

**Comment by Caspervw** [comment]
@mkendall07  Creating a test-page for this one is kinda hard. What needs to happen:

- Request an advertisement and make sure we get a bid from someone (ex. AppNexus)
- This bid has to lose to the backfill, so that it ends up in the client-side cache for TTL time.
- Refresh and receive another bid from someone (ex. AppNexus), this needs to be the same provider.

This will cause the cached bid and the new bid to be merged into one targeting-group for DFP, so that DFP can pick the winning bid. 

I think the concatenation happens on https://github.com/prebid/Prebid.js/blob/master/src/targeting.js#L113, although there it add a ", " and not a ",".

**Comment by Caspervw** [comment]
Alright, so I’m going to close this one, because I think it’s solved and works as expected. 

What happened is that we were using the “adserverTargeting” to change the prices. This turned everything we defined (ex. hb_adId_xxx) into a “custom targeting”, which are merged by the “mergeAdServerTargeting” in targeting.js. This caused the concatenation of adId values. 

Solution: We removed the "adserverTargeting" and now change our prices in bidCpmAdjustment.

**Comment by Caspervw** [comment]
Last comment. Does http://prebid.org/dev-docs/examples/custom-price-bucket.html contain the same problem? Do the “custom_bid_price_key”’s get merged into 0.10,0.20,0.05 etc etc? If so, Prebid might behave erratically if run with refreshing turned on.

Good luck with this library! :+1:

**Comment by pm-harshad-mane** [comment]
Hello @mkendall07 and @Caspervw ,
Where can I find documentation about the bid-caching feature ?
